File: sys/petscOptionsGetString.m
function [str, found, errCode, toplevel] = petscOptionsGetString(opts, pre, name)
% Gets the string value for a particular option in the database. 
%
%   [str, found, errCode, toplevel] = petscOptionsGetString(opts, pre, name)
%   obtains a string in the data base. The flag found is PETSC_TRUE
%   if the attribute was found. The strings must be null-terminated.
%
% SEE ALSO: petscOptionsInsertString, petscOptionsHasName, 
%           PetscOptionsGetInt, PetscOptionsGetString
%
% PETSc C interface:
%   PetscErrorCode  PetscOptionsGetString(PetscOptions options,const char pre[],const char name[],char string[],size_t len,PetscBool  *set)

% http://www.mcs.anl.gov/petsc/petsc-current/docs/manualpages/Sys/PetscOptionsGetString.html

%#codegen -args {PetscOptions, coder.typeof(char(0), [1,inf]), coder.typeof(char(0), [1,inf])}

errCode = int32(-1);

if ~coder.target('MATLAB')
    toplevel = nargout>3;
    if ~isempty(pre) && pre(end) && (toplevel || m2c_debug)
        m2c_error('MPETSc:petscOptionsGetString:InputError', ...
            'Argument pre must be a null-terminated string.')
    end
    if ~isempty(name) && name(end) && (toplevel || m2c_debug)
        m2c_error('MPETSc:petscOptionsGetString:InputError', ...
            'Argument name must be a null-terminated character string.')
    end
    
    b_flag = coder.opaque('PetscBool');
    
    str0 = char(zeros(1, 21));
    errCode = coder.ceval('PetscOptionsGetString', PetscOptions(opts), ...
        coder.rref(pre), coder.rref(name), coder.wref(str0), int32(20), coder.wref(b_flag));

    found = int32(0); %#ok<NASGU>
    found = coder.ceval('(int)', b_flag);
    
    str = '';
    for i=1:21
        if str0(i)==0
            str = str0(1:i);
            break;
        end
    end
    
    if errCode && (toplevel || m2c_debug)
        m2c_error('petsc:RuntimeError', 'PetscOptionsGetString returned error code %d\n', errCode)
    end
end

end

File: sys/codegen/lib/petscOptionsGetString/petscOptionsGetString.c
#include "petscOptionsGetString.h"
#include "petscOptionsGetString_types.h"
#include "m2c.h"
#include "petsc4m.h"
#include <string.h>

static void b_m2c_error(void);

static void c_m2c_error(int varargin_3);

static void m2c_error(void);

static void b_m2c_error(void)
{
  const char *fmt;
  const char *msgid;
  msgid = "Petsc4m:petscOptionsGetString:InputError";
  fmt = "Argument name must be a null-terminated character string.";
  M2C_error(msgid, fmt);
}

static void c_m2c_error(int varargin_3)
{
  const char *fmt;
  const char *msgid;
  msgid = "petsc:RuntimeError";
  fmt = "PetscOptionsGetString returned error code %d\n";
  M2C_error(msgid, fmt, varargin_3);
}

static void m2c_error(void)
{
  const char *fmt;
  const char *msgid;
  msgid = "Petsc4m:petscOptionsGetString:InputError";
  fmt = "Argument pre must be a null-terminated string.";
  M2C_error(msgid, fmt);
}

void petscOptionsGetString(const emxArray_char_T *pre,
                           const emxArray_char_T *name, char str_data[],
                           int str_size[2], int *found, int *errCode,
                           boolean_T *toplevel)
{
  PetscBool b_flag;
  PetscOptions obj;
  int i;
  char str0[21];
  const char *name_data;
  const char *pre_data;
  boolean_T exitg1;
  name_data = name->data;
  pre_data = pre->data;
  *toplevel = true;
  if ((pre->size[1] != 0) && (pre_data[pre->size[1] - 1] != '\x00')) {
    m2c_error();
  }
  if ((name->size[1] != 0) && (name_data[name->size[1] - 1] != '\x00')) {
    b_m2c_error();
  }
  obj = NULL;
  *errCode = PetscOptionsGetString(obj, &pre_data[0], &name_data[0], &str0[0],
                                   20, &b_flag);
  *found = (int)(b_flag);
  str_size[0] = 1;
  str_size[1] = 0;
  i = 0;
  exitg1 = false;
  while ((!exitg1) && (i < 21)) {
    if ((unsigned char)str0[i] == 0) {
      str_size[0] = 1;
      str_size[1] = i + 1;
      memcpy(&str_data[0], &str0[0], (i + 1) * sizeof(char));
      exitg1 = true;
    } else {
      i++;
    }
  }
  if (*errCode != 0) {
    c_m2c_error(*errCode);
  }
}

void petscOptionsGetString_initialize(void)
{
}

void petscOptionsGetString_terminate(void)
{
}
